### ISSUE TITLE
feat: Support for 'discovery_type' in ApisixUpstream

### ISSUE THREAD

**Issue by joyyir** [apache/apisix-ingress-controller#948]
Hello,

When integrating with service discovery, `discovery_type` is specified in the upstream settings as described in the following document: https://apisix.apache.org/docs/apisix/discovery/#upstream-setting

However, there is no `discovery_type` field in the `ApisixUpstream` spec, so to use the service discovery integration I need to register upstream with the dashboard or Admin API.

So I propose to add `discovery_type` in `ApisixUpstream` spec.

Thanks.

**Comment by tao12345666333** [comment]
Can you describe your specific use case?  I would like to know what service do you need to integrate with

**Comment by joyyir** [comment]
I want to integrate with `Eureka`. I added the upstream setting by dashboard. I found that the upstream setting cannot be defined as `ApisixStream` because of `discovery_type`.

Here is the upstream setting:
```yaml
name: my-svc-upstream
type: ewma
scheme: http
discovery_type: eureka
pass_host: pass
service_name: MY-SVC
```

Here is the route setting:
```yaml
uri: /my-svc/*
name: my-svc-route
host: www.my-host.com
upstream_id: '401325612930696050'
```

**Comment by tao12345666333** [comment]
I understand your situation.
But the current APISIX Ingress controller is designed to automatically discover changes to resources in Kubernetes.
We need some discussion about this feature.

**Comment by tokers** [comment]
Only support the discovery type in ApisixUpstream cannot satisfy your needs since APISIX Ingress Controller only watches service from Kubernetes, and ApisixUpstream will be in effective only if there is a Kubernetes serivce which name is same to the ApisixUpstream.

**Comment by tao12345666333** [comment]
After thinking about it, we actually have a relatively simple way to achieve it.

* Configure multiple discovery types in APISIX
* Extend ApisixUpstream resource to allow configuration of discovery_type
* If the discovery_type is found in the ApisixUpstream resource, it will no longer be parsed, and the corresponding Upstream will be created in APISIX

**Comment by Gallardot** [comment]
The Integration service discovery registry in APISIX is a very attractive feature.  Support for this feature will help expand the usage scenarios for apisix-ingress-controller.

I want to have a try.

**Comment by tao12345666333** [comment]
@Gallardot sorry for delay.  Can you prepare a proposal first? Contains general implementation content, etc. Thanks

**Comment by tao12345666333** [comment]
I've implemented that functionality in #1465 .
It will be available in v1.6 release.
thanks

e.g.
```
apiVersion: apisix.apache.org/v2
kind: ApisixUpstream
metadata:
  name: foo
spec:
  discovery:
    type: dns #consul/nacos/ and others
    serviceName: bar
```

**Comment by yujinchoi-94** [comment]
@tao12345666333 
Hi,
`discovery` related fields seem missing in [ApisixUpstream reference](https://apisix.apache.org/docs/ingress-controller/references/apisix_upstream/)
It would be good if a spec for a `discovery` related fields are added.
